include/swift/AST/Expr.h

```diff
@@ -3768,7 +3766,25 @@ class SerializedAbstractClosureExpr : public SerializedLocalDeclContext {
 ///     { [weak c] (a : Int) -> Int in a + c!.getFoo() }
 /// \endcode
 class ClosureExpr : public AbstractClosureExpr {
+public:
+  enum class BodyState {
+    /// The body was parsed, but not ready for type checking because
+    /// the closure parameters haven't been type checked.
+    Parsed,
 
+    /// The type of the closure itself was type checked. But the body has not
+    /// been type checked yet.
+    ReadyForTypeChecking,
+
+    /// The body was typechecked with the enclosing closure.
+    /// i.e. single expression closure or function builder closure.
+    TypeCheckedAtOnce,
```

Comment:
Nit: probably meant to be `TypeCheckedOnce`.

Reply:
Renamed to `TypeCheckedWithSignature` to clarify the meaning.